scripts/tests/AUTH/test_3b_key_migration_execute.py

```
@@ -131,10 +130,25 @@ def create_test_user(api_base: str, user_data: dict, admin_token: str) -> dict |
-    cmd = f"docker compose exec -T backend sqlite3 -json {shlex.quote(db_path)} '{escaped_sql}'"
-    result = subprocess.run(cmd, shell=True, capture_output=True, text=True, cwd=REPO_ROOT)
+    """
+    Run read-only SQL inside Docker container and return output.
+
+    Security: Uses stdin to pass SQL (avoids shell injection), and
+    validates that only SELECT statements are allowed.
+    """
+    # Validate: only allow SELECT statements (defense-in-depth for test helper)
+    sql_normalized = sql.strip().upper()
+    if not sql_normalized.startswith("SELECT"):
+        raise ValueError(f"run_docker_sql only allows SELECT statements, got: {sql[:50]}")
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)